crypto/rsa/rsa_gen.c

```diff
@@ -85,62 +118,148 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
+        } else if (i == 1) {
+            prime = rsa->q;
+        } else {
+            pinfo = sk_RSA_PRIME_INFO_value(prime_infos, i - 2);
+            prime = pinfo->r;
+        }
+
+        for (;;) {
+ redo:
+            if (!BN_generate_prime_ex(prime, bitsr[i] + adj, 0, NULL, NULL, cb))
+                goto err;
+            /*
+             * prime should not be equal to p, q, r_3...
+             * (those primes prior to this one)
+             */
+            {
+                int j = 0;
+                BIGNUM *prev_prime = NULL;
+
+                for (j = 0; j < i; j++) {
+                    if (j == 0)
+                        prev_prime = rsa->p;
+                    else if (j == 1)
+                        prev_prime = rsa->q;
+                    else
+                        prev_prime = sk_RSA_PRIME_INFO_value(prime_infos, j - 2)->r;
+
+                    if (!BN_cmp(prime, prev_prime)) {
+                        goto redo;
+                    }
+                }
+            }
+            if (!BN_sub(r2, prime, BN_value_one()))
+                goto err;
+            if (!BN_gcd(r1, r2, rsa->e, ctx))
+                goto err;
+            if (BN_is_one(r1))
+                break;
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+        }
+        if (primes == RSA_DEFAULT_PRIME_NUM && i == 1) {
+            if (BN_cmp(rsa->p, rsa->q) < 0) {
+                tmp = rsa->p;
+                rsa->p = rsa->q;
+                rsa->q = tmp;
+            }
+        }
+        bitse += bitsr[i];
+
+        /* calculate n immediately to see if it's sufficient */
+        if (i == 1) {
+            /* we get at least 2 primes */
+            if (!BN_mul(r1, rsa->p, rsa->q, ctx))
+                goto err;
+        } else if (i != 0) {
+            /* modulus n = p * q * r_3 * r_4 ... */
+            if (!BN_mul(r1, rsa->n, prime, ctx))
+                goto err;
+        } else {
+            /* i == 0, do nothing */
+            if (!BN_GENCB_call(cb, 3, i))
                 goto err;
-        } while (BN_cmp(rsa->p, rsa->q) == 0);
-        if (!BN_sub(r2, rsa->q, BN_value_one()))
+            continue;
+        }
+        nl = BN_num_bits(r1);
+        /* if n is not long enough, re-generate the last prime */
+        if (nl != bitse) {
+            /* increase the bits of last prime */
+            adj += bitse - nl;
+            bitse -= bitsr[i];
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+            goto redo;
+        }
+        /* save product of primes for further use, for multi-prime only */
+        if (i > 1 && BN_copy(pinfo->pp, rsa->n) == NULL)
             goto err;
-        if (!BN_gcd(r1, r2, rsa->e, ctx))
+        if (BN_copy(rsa->n, r1) == NULL)
             goto err;
-        if (BN_is_one(r1))
-            break;
-        if (!BN_GENCB_call(cb, 2, n++))
+        if (!BN_GENCB_call(cb, 3, i))
             goto err;
+        adj = 0;
```

Comment:
This is very counter-intuitive placement for zeroing. [In practical terms it means that I made wrong assumption when reading code and was wondering about it.] Line 157, if not directly as `for (adj = 0;;)` at line 158 would be much better spot. [In which case initial assignment at declaration time is rendered redundant. One can argue that it would be more appropriate to declare it in closer scope, i.e. in the loop at line 148.]

Reply:
On related note, it seems that for 4 and less primes, it makes sense to not adjust factor bit length. Because one has a lot to win by adhering to 1024-bit factors in rsa3072 and rsa4096. First of all probability that you'd need adjustment is rather low for such small number of primes. Which in turn means that probability that you'd have to adjust back is rather high. Or in other words you'd be better off if you simply try another factor of *same* bit length. On the other hand in unlike event of one 1025-bit factor you can observe quite a performance hit. For example on Haswell I've measured 29% loss in rsa3072 and 23% in rsa4096.

Reply:
Setting limit for adjustment would also eliminate formal doubt about equivalence to two-prime code. I mean suggested code is equivalent, but only because probability of need for adjustment is 0 in two-prime case. [Was it actually recognized?]

And once again, comparison between p and q can and should be taken out of loop and performed unconditionally for all number of primes.

Reply:
> This is very counter-intuitive placement for zeroing

This is very subjective and I could also say it's very intuitive because this place is where we just get a valid prime and just multiply it with `rsa->n` - the modules. It's the place ready for generating next prime, so the variable is zero'ed there.

> First of all probability that you'd need adjustment is rather low for such small number of primes.

If so, then why do we need to treat 4 and less primes key specially? I mean, if the probability of adjustment is very low for those keys, then the probability for generating a 1025-bit prime is also very low...

> Which in turn means that probability that you'd have to adjust back is rather high. 

I don't get this, what to adjust back?

Reply:
> > Which in turn means that probability that you'd have to adjust back is rather high.

> I don't get this, what to adjust back?

Let's say you generate 3rd 1024-bit factor. There is slim chance that product will be 3071 bits. As it stands now you choose to generate 1025-bit factor, but chances are even higher that this time product will be 3073 bits. In which case you go back attempting 1024-bit factor. You actually have better chances just attempting another 1024-bit factor, instead of going for 1025 and backing to 1024 that is. Yes, there is slim chance that 1025-bit factor would work, *but* it will cost you performance, Going for 4th prime. Chance that product will be 4095 bits is higher, but still rather low, and so is chance that it will be 4097 bits with last 1025-bit factor. I reckon that performance gain is worth retrying same length factor. One should be able to estimate when it doesn't pay off to attempt same length, but limit of 4 does work empirically and is a practical case for rsa4096, i.e. likely to be used in practice.

Reply:
> Chance that product will be 4095 bits is higher, but still rather low, and so is chance that it will be 4097 bits with last 1025-bit factor.

Just in case, I meant that chances for 4097 is lower than for getting 3073 with 3rd factor, but it's still low enough to attempt same length.

Reply:
> > This is very counter-intuitive placement for zeroing

> This is very subjective and I could also say it's very intuitive

"Could" or "do"? Note that what you are saying is that setting up initial value for *inner* loop at the end of previous iteration of outer loop is more intuitive...

Reply:
Hmmm, this is interesting. So your suggestion is to limit adjustments for 3 or 4 primes key with more than 3072 bit?

Reply:
From my point of view, what you suggested has two points for not adjusting prime length if number of prime is less and equal than 4:

1. reduce the chance of 'adjust-back' when generating the prime
2. avoid performance issues if prime is adjusted to something like 1025 bits

For the first reason, it's not sufficient to make a change (adding limitations for example), since the adjust-back works fine and the probability is low so there has nothing to do with performance (well, might have *little* thing to do with performance of key generation...)

For the second reason, it *might* be worth to set a limit to avoid performance issue. But it also feasible to leave this choice to the user, say if we and a 'flag' here to turn-off the adjustment logic and let the users to set it if they need...

Reply:
First of all suggestion is to not do adjustments for 2, 3 and 4 factors [not just 3 and 4]. And cases of 2 and 3 are not actually specific to any [public] key length. In case of 2 goal is to ensure formal equivalence with original two-prime code. [I mean original didn't do any comparison, so new code shouldn't either.] In case of 3 you actually have better chances retrying same key length, which means that *on average* you'll generate 3rd factor faster. And only case of 4 has some bearing to key length. Retrying same factor length will avoid performance loss for lengths that are power of 2(\*). *On average* key generation might be a little bit slower (because once in a while it might take two tries with same length instead of one), but it's worth it. Yes, one can say that then one could avoid adjustment only in cases when [public] key length is power of 2, but suggestion is to let it slide to not over-complicate things. 

(\*) It just so happens that rsa4096 is likely to be one that benefits most in current combination of code paths, but it does benefit *all* powers of 2.

Reply:
> leave this choice to the user

The choice (and the is choice really only in case of 4) is overwhelmingly sensible for absolute majority of users. I'd say that it's so sensible that those who insist on having one can as well be told to modify source code.

Reply:
I did some tests, and seems 4-prime is also OK for not looping a lot if there is no adjustment, so I simply let 2, 3 ,4 not adjust no matter how long the length of key is...

Reply:
I agree with @dot-asm here, please move the `adj = 0;` immediately after the line `for (i = 0; i < primes; i++) {`

Reply:
Okay, I guess I will move that above since most people consider this as proper.